baseapp/msg_service_router.go

```diff
@@ -0,0 +1,83 @@
+package baseapp
+
+import (
+	"fmt"
+
+	gogogrpc "github.com/gogo/protobuf/grpc"
+	"google.golang.org/grpc"
+
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// MsgServiceRouter routes Msg Service fully-qualified service methods to their
+// handler.
+type MsgServiceRouter struct {
+	interfaceRegistry codectypes.InterfaceRegistry
+	routes            map[string]sdk.Handler
+}
+
+var _ gogogrpc.Server = &MsgServiceRouter{}
+
+// NewMsgServiceRouter creates a new MsgServiceRouter.
+func NewMsgServiceRouter() *MsgServiceRouter {
+	return &MsgServiceRouter{
+		routes: map[string]sdk.Handler{},
+	}
+}
+
+// sdk.Handler defines a function type which handles Msg service message.
+// It's similar to sdk.Handler, but with simplified version of `Msg`, without
+// `Route()`, `Type()` and `GetSignBytes()`.
+type MsgServiceHandler = func(ctx sdk.Context, msgRequest sdk.MsgRequest) (*sdk.Result, error)
+
+// Route returns the sdk.Handler for a given query route path or nil
+// if not found.
+func (msr *MsgServiceRouter) Route(path string) sdk.Handler {
+	handler, found := msr.routes[path]
+	if !found {
+		return nil
+	}
+
+	return handler
+}
+
+// RegisterService implements the gRPC Server.RegisterService method. sd is a gRPC
+// service description, handler is an object which implements that gRPC service.
+func (msr *MsgServiceRouter) RegisterService(sd *grpc.ServiceDesc, handler interface{}) {
+	// adds a top-level query handler based on the gRPC service name
+	for _, method := range sd.Methods {
+		fqMethod := fmt.Sprintf("/%s/%s", sd.ServiceName, method.MethodName)
+		methodHandler := method.Handler
+
+		msr.routes[fqMethod] = func(ctx sdk.Context, _ sdk.Msg) (*sdk.Result, error) {
```

Comment:
It seems weird to me that I don't use the sdk.Msg argument here.

Reply:
We don't need `sdk.Handler` here. Just use a function type that makes sense. It should probably be `func(ctx sdk.Context, request []byte) (*sdk.Result, error)`. And that should be the return type for `Route` above.

Reply:
Got it. The reason I went with `sdk.Msg` was to match the signature required by baseapp in `app#runMsgs`. I remove that too.